src/make_features/extract_subject_verb_object_from_content_titles.py

```diff
@@ -0,0 +1,55 @@
+import pandas as pd
+from src.make_features.subject_verb_object.content import Page
+import spacy
+import json
+
+def get_verbs_objects(pages, nlp):
+    objects = {}
+    verbs = {}
+    len = all_content_items.shape[0]
+    for index, page in enumerate(pages):
+        print(f"{index} of {len}")
+        for title in page.titles(nlp):
+            for triple in title.subject_object_triples():
+                triple_object = triple.cypher_object()
+                triple_verb = triple.cypher_verb()
+                if triple_object not in objects:
+                    objects[triple_object] = []
+                objects[triple_object].append([triple_verb, page.base_path(), title.title])
+                if triple_verb not in verbs:
+                    verbs[triple_verb] = []
+                verbs[triple_verb].append([triple_object, page.base_path(), title.title])
+    print("Found all SVOs, making them unique")
+    verbs = find_unique_entries(verbs)
+    objects = find_unique_entries(objects)
+    return verbs, objects
+
+def find_unique_entries(verbs_or_objects):
+    unique_entries = {}
+    for key, value in verbs_or_objects.items():
+        unique_values = []
+        for item in value:
+            found = False
+            for unique_item in unique_values:
+                if unique_item[0] == item[0]:
+                    found = True
+            if not found:
+                unique_values.append(item)
+        unique_entries[key] = unique_values
+    return unique_entries
```

Comment:
So (i think) this leaves us with a dictionary of (e.g.) objects and a set of all the verbs you can do to that object, but this throws away valuable information - the counts of co-occurences between verbs and objects/objects verbs. ie if someone types feedback, we might wanna estimate the probability of each verb given that. e.g.  p(give | feedback) ~ 0.9, and p(delivered | feedback) ~ 0.1. Might be more useful downstream to associate each object with a distribution of verbs given that object and vice versa.

Think we need to think through what probability are we trying to estimate here in general, fundamentally how does knowing a word is a verb/object in a title improve our estimate of:

p(user wants page | query)

Reply:
This is a really really good point. Not only that, but I'd been wondering what to do about pages like "Don't rubbish your passport" where it ends up thinking that you can rubbish->passport (which you can, but it's hardly a key feature) - if it's only ever seen once then we can say that it might not really be a thing - or that it's much less common then applying or renewing - ie p(verb is a core feature of that entity|number of co-occurances)

> Think we need to think through what probability are we trying to estimate here in general, fundamentally how does knowing a word is a verb/object in a title improve our estimate of:
>
> p(user wants page | query)

My theory is that if a verb applies to an entity in a title then the likelihood that the page is about that entity is increased. To continue the Bayesian analogy - my theory is
p(page is about an entity| entity is object of verb in title) > p(page is about an entity| entity is in title)
And that
p(user wants page | entity in query is the object of an SVO triple) > p(user wants page | entity in query is mentioned in the title)
This is also based on the prior that we know that people generally make short queries (I think you said the average was 2 words)
Thus, if an entity in a query is the object of a verb means it's more 'present' in a page, for example for the entity 'council', there are pages like
"Find your local council"
and
"Apply to adopt a child through your local council"
Where in the first council is the object of a verb and in the second it isn't. The first page is primarily about councils while the second is only somewhat about councils. Given that people make short queries, someone entering 'council' it's more likely that they want to do something directly in relation to the council and if they wanted to adopt a child they'd probably be entering something more about that. 
Certainly, as a human, if someone told you "council" and you had to guess what they wanted, you probably wouldn't think it'd be something about adopting - so it's modelling that. As the user enters more info though, those probabilities will obvs have to change
I'll certainly, happily, concede that it's possible that they could be wanting something about councils and adoption but given all we know, the alternatives are more likely. In the future, I'd like to use other relationships between words to understand that although council is less important in the second title, it is still there and that someone typing 'council adoption' should get the second page even though it's not the object - this is the simplest and easiest single method (that I can think of) of trying to make something more sophisticated than simply using word presence to make suggestions,.

Reply:
**So the simplest possible implementation for recommending titles.**

`p(title | query_token) = (p(query_token | title) * p(title))`

So how do we figure out p(query_token | title).

We could just say its a frequency count of each word in a title / title length. Really we want the probability of the user using that token given they want that page, but this will do!)

So say we the user types 'apply'. Let's imagine:

`p(apply for a passport | 'apply') = 0.25 * 0.6
`
`p(licence to apply paint | 'apply') = 0.25 * 0.4
`
passport wins!

When does knowing 'apply' is a verb, help? 

To use your example. say find your local council and applying to adopt were equally popular.

`p(Find your local council | council) = 0.25 * 0.5
`
`p(Apply to adopt a child through your local council | council) = 0.11 * 0.5
`
(note this works!!!)

But youre saying knowing something is the main object means we should revise our estimate of p(token).  If we know something is a the main object, users are gonna use that token more, because it has more to do with the actual content. So we could just add an (arbitrary) prior to our p(token) by adding a pseudo count to the words that are objects/verbs in our titles and recalc the probabilities.

ie instead of find: 1, your:1, local:1, council:1

we cheekily do find:10, your:1, local:1, council:10

(note this is a rough interpretation of p(word | word_dis) * p (word_dis))

So we end up with something like

`p(Find your local council | council) = 0.8333 * 0.5
`
We can maybe set our prior by considering how much more likely subject/verbs are compared to other tokens.

Reply:
Interesting :+1: Would be good to take a further look into this at some point